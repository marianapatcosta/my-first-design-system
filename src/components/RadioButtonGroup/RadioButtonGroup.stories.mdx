import { Meta } from '@storybook/addon-docs/blocks';
import { RadioButtonGroup } from './RadioButtonGroup';
import { ORIENTATION } from '../../constants';

<Meta title='Design System|RadioButtonGroup' component={RadioButtonGroup} />

# RadioButtonGroup

RAdio button group components; groups radio button elements

## Usage

Use radio button group component to display a predefine group of options, represented by RadioButton components,
from which the user should choose only one.

- Default RadioButtonGroup:

  <Story name='Default'>
    <RadioButtonGroup />
  </Story>

- Vertical RadioButtonGroup:
  <Story name='Vertical'>
    <RadioButtonGroup
      items={[
        { label: 'option 1', name: 'name 1' },
        { label: 'option 2', name: 'name 2' },
        { label: 'option 3', name: 'name 3' },
      ]}
      orientation={ORIENTATION.VERTICAL}
    />
  </Story>

## Other important notes

not applicable.
