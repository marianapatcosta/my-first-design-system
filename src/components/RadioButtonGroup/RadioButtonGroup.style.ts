import styled from 'styled-components';
import { baseStyles } from '../../styles';
import { RadioButton } from '../RadioButton/RadioButton';

export const StyledRadioButtonGroup = styled.div`
  ${baseStyles}
  display: flex;
  flex-direction: column;
`;

export const StyledRadioButtonGroupTitle = styled.h3`
  font-size: 110%;
  font-weight: normal;
  margin-bottom: 1rem;
`;

export const StyledRadioButtonsContainer = styled.div<{ isVertical?: boolean }>`
  display: flex;
  > * {
    margin-right: 1rem;
  }
  ${({ isVertical }) =>
    isVertical &&
    `flex-direction: column; > * {
     margin-bottom: 0.5rem; margin-right: 0;}`}
`;
