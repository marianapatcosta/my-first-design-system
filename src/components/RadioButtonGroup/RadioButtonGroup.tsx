import React, { CSSProperties, ChangeEvent } from 'react';
import { ORIENTATION } from '../../constants';
import { RadioButton } from '../RadioButton/RadioButton';
import {
  StyledRadioButtonGroup,
  StyledRadioButtonGroupTitle,
  StyledRadioButtonsContainer,
} from './RadioButtonGroup.style';

type orientationTypes = ORIENTATION;

interface RadioButtonItem {
  label: string;
  id?: string;
  name?: string;
}

export interface RadioButtonGroupProps {
  /**
   * element id
   */
  id?: string;
  /**
   * additional css class
   */
  className?: string;
  /**
   * additional style
   */
  style?: CSSProperties;
  /**
   * radio button group's title
   */
  title?: string;
  /**
   * Are radio buttons disabled?
   */
  disabled?: boolean;
  /**
   * radio button group items' orientation
   */
  orientation?: orientationTypes;
  /**
   * additional css class
   */
  radioButtonClassName?: string;
  /**
   * items to diaplay in radiobutton
   */
  items: Array<RadioButtonItem> | Array<string>;
  /**
   * items to diaplay in radiobutton
   */
  selectedItem: string | Array<RadioButtonItem>;
  /**
   * Change handler
   */
  onChange: (event?: ChangeEvent<HTMLInputElement>) => void;
}

export const RadioButtonGroup: React.FC<RadioButtonGroupProps> = ({
  items,
  title,
  disabled,
  radioButtonClassName,
  selectedItem,
  orientation,
  onChange,
  ...props
}) => {
  return (
    <StyledRadioButtonGroup {...props} role='radiogroup'>
      {title && (
        <StyledRadioButtonGroupTitle>{title}</StyledRadioButtonGroupTitle>
      )}
      <StyledRadioButtonsContainer
        isVertical={orientation === ORIENTATION.VERTICAL}
      >
        {items.map((item: any) => (
          <RadioButton
            key={`radio-button-${item.label || item}`}
            disabled={disabled}
            id={item.id}
            name={item.name}
            className={radioButtonClassName}
            checked={
              !!item.label ? selectedItem === item.label : selectedItem === item
            }
            value={item.label || item}
            label={item.label || item}
            onChange={onChange}
          />
        ))}
      </StyledRadioButtonsContainer>
    </StyledRadioButtonGroup>
  );
};

RadioButtonGroup.defaultProps = {
  id: '',
  className: '',
  style: {},
  title: '',
  orientation: ORIENTATION.LANDSCAPE,
  radioButtonClassName: '',
  items: [],
  selectedItem: '',
  onChange: () => null,
};
