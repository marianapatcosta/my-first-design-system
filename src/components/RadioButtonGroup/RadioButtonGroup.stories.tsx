import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { RadioButtonGroup, RadioButtonGroupProps } from './RadioButtonGroup';
import { ORIENTATION } from '../../constants';

export default {
  title: 'RadioButtonGroup',
  component: RadioButtonGroup,
  args: {
    title: 'radio group',
    items: [
      { label: 'option 1', name: 'name 1' },
      { label: 'option 2', name: 'name 2' },
      { label: 'option 3', name: 'name 3' },
    ],
  },
} as Meta;

const Template: Story<RadioButtonGroupProps> = (args) => (
  <RadioButtonGroup {...args} />
);

export const Default = Template.bind({});
Default.args = {};

export const Vertical = Template.bind({});
Vertical.args = { orientation: ORIENTATION.VERTICAL };
