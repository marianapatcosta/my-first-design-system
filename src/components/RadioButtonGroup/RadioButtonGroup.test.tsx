import React from 'react';
import { render } from '@testing-library/react';
import { RadioButtonGroup } from './RadioButtonGroup';

describe('RadioButtonGroup', () => {
  const defaultProps = {
    title: 'radio group',
    items: [
      { label: 'option 1', name: 'name 1' },
      { label: 'option 2', name: 'name 2' },
      { label: 'option 3', name: 'name 3' },
    ],
    selectedItem: '',
    onChange: () => null,
  };
  test('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(<RadioButtonGroup {...defaultProps} />);

    expect(firstChild).toMatchSnapshot();
  });

  test('should return null on change if onChange prop is not passed', () => {
    const result =
      RadioButtonGroup.defaultProps?.onChange &&
      RadioButtonGroup.defaultProps.onChange();
    expect(result).toBeNull();
  });

  test('should display title, if `title`prop is passed', () => {
    const { getByText } = render(<RadioButtonGroup {...defaultProps} />);

    getByText(/radio group/);
  });

  test('should pass props to RadioButton component', () => {
    const { getByLabelText } = render(<RadioButtonGroup {...defaultProps} />);

    const radioButton = getByLabelText(/option 1/);
    expect(radioButton).toBeInTheDocument();
  });
});
