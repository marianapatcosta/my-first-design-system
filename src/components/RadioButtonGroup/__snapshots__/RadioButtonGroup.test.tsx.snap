// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButtonGroup renders without errors and matches snapshot 1`] = `
<div
  class="sc-hKgILt hoZpjz"
  id=""
  role="radiogroup"
>
  <h3
    class="sc-eCssSg iBHIVn"
  >
    radio group
  </h3>
  <div
    class="sc-jSgupP jTPNBf"
  >
    <label
      aria-checked="false"
      class="sc-bdfBwQ ivjLLh"
    >
      <input
        class="sc-dlfnbm bGKYVm"
        data-testid="radio-button"
        name="name 1"
        type="radio"
        value="option 1"
      />
      <span
        class="sc-gsTCUz jzeJuN"
      />
      option 1
    </label>
    <label
      aria-checked="false"
      class="sc-bdfBwQ ivjLLh"
    >
      <input
        class="sc-dlfnbm bGKYVm"
        data-testid="radio-button"
        name="name 2"
        type="radio"
        value="option 2"
      />
      <span
        class="sc-gsTCUz jzeJuN"
      />
      option 2
    </label>
    <label
      aria-checked="false"
      class="sc-bdfBwQ ivjLLh"
    >
      <input
        class="sc-dlfnbm bGKYVm"
        data-testid="radio-button"
        name="name 3"
        type="radio"
        value="option 3"
      />
      <span
        class="sc-gsTCUz jzeJuN"
      />
      option 3
    </label>
  </div>
</div>
`;
